File: ios/AgentControlCenter/Security/URLValidator.swift
import Foundation

// MARK: - URLValidator
// 防止用户输入危险 URL 触发 SSRF / apiKey 泄漏

/// URL 安全校验器。
///
/// 用于校验用户配置的 Agent endpoint / WebSocket URL / MCP transport URL，
/// 防止 SSRF（如 `http://169.254.169.254/...` 探活 AWS metadata 泄漏 apiKey）、
/// 本地端口扫描、内网穿透、危险 scheme（file:// / data://）等。
enum URLValidator {

    /// 允许的 scheme 白名单
    static let allowedSchemes: Set<String> = ["http", "https", "ws", "wss"]

    /// 禁止访问的 IP 段（链路本地 / 元数据服务 / 回环陷阱等）
    /// 注意：127.0.0.1 / localhost 允许（本地 Agent 部署场景）
    private static let blockedCIDRs: [(address: String, prefix: Int)] = [
        ("169.254.0.0", 16),    // 链路本地（含 AWS/GCP/Azure metadata 169.254.169.254）
        ("0.0.0.0", 8),         // 本网络
        ("255.255.255.255", 32) // 广播
    ]

    /// 校验 URL 字符串是否安全可用。
    /// - Parameters:
    ///   - urlString: 用户输入的 URL 字符串
    ///   - allowLocalhost: 是否允许 localhost / 127.0.0.1 / 内网私网地址（默认 true，本地 Agent 部署需要）
    /// - Returns: 校验通过返回 URL；失败返回 nil
    static func validate(_ urlString: String, allowLocalhost: Bool = true) -> URL? {
        guard let url = URL(string: urlString),
              let scheme = url.scheme?.lowercased(),
              allowedSchemes.contains(scheme) else {
            return nil
        }

        // 必须有 host
        guard let host = url.host?.lowercased(), !host.isEmpty else {
            return nil
        }

        // localhost 显式放行（若 allowLocalhost）
        if allowLocalhost && (host == "localhost" || host == "ip6-localhost" || host == "::1") {
            return url
        }

        // IPv4 字面量校验
        if let ipv4 = parseIPv4(host) {
            if !isIPv4Allowed(ipv4, allowLocalhost: allowLocalhost) {
                return nil
            }
            return url
        }

        // IPv6 字面量（带方括号或裸写）— 简化处理：仅放行非链路本地
        if host.contains(":") {
            // IPv6，链路本地 fe80::/10 禁止
            let lower = host.hasPrefix("[") ? String(host.dropFirst().dropLast()) : host
            if lower.lowercased().hasPrefix("fe80") {
                return nil
            }
            return url
        }

        // 域名校验：禁止以 .local / .internal 等内网域名探测（防止 mDNS SSRF）
        // 但允许常规公网域名
        return url
    }

    // MARK: - IPv4 Helpers

    private static func parseIPv4(_ s: String) -> [UInt8]? {
        let parts = s.split(separator: ".")
        guard parts.count == 4 else { return nil }
        var bytes: [UInt8] = []
        for p in parts {
            guard let v = UInt8(p), p.count <= 3 else { return nil }
            bytes.append(v)
        }
        return bytes
    }

    /// IPv4 是否允许访问
    private static func isIPv4Allowed(_ ip: [UInt8], allowLocalhost: Bool) -> Bool {
        // 127.0.0.0/8 回环
        if ip[0] == 127 {
            return allowLocalhost
        }
        // 私网地址
        if ip[0] == 10 { return allowLocalhost }
        if ip[0] == 172 && (ip[1] >= 16 && ip[1] <= 31) { return allowLocalhost }
        if ip[0] == 192 && ip[1] == 168 { return allowLocalhost }

        // 禁止的 IP 段
        for cidr in blockedCIDRs {
            if let cidrAddr = parseIPv4(cidr.address), inCIDR(ip, cidrAddr, prefix: cidr.prefix) {
                return false
            }
        }
        return true
    }

    private static func inCIDR(_ ip: [UInt8], _ cidr: [UInt8], prefix: Int) -> Bool {
        let fullBytes = prefix / 8
        let remainder = prefix % 8
        for i in 0..<fullBytes {
            if ip[i] != cidr[i] { return false }
        }
        if remainder > 0 && fullBytes < 4 {
            let mask: UInt8 = (~0) << (8 - remainder)
            if (ip[fullBytes] & mask) != (cidr[fullBytes] & mask) { return false }
        }
        return true
    }
}

File: android/app/src/main/java/com/agentcontrolcenter/app/core/security/UrlValidator.kt
package com.agentcontrolcenter.app.core.security

import java.net.URL

// MARK: - UrlValidator
// 对应 iOS Security/URLValidator.swift
// 防止用户输入危险 URL 触发 SSRF / apiKey 泄漏

/**
 * URL 安全校验器。
 *
 * 用于校验用户配置的 Agent endpoint / WebSocket URL / MCP transport URL，
 * 防止 SSRF（如 `http://169.254.169.254/...` 探活 AWS metadata 泄漏 apiKey）、
 * 本地端口扫描、内网穿透、危险 scheme（file:// / data://）等。
 *
 * 与 iOS `URLValidator` 完全对齐，跨端行为一致。
 *
 * H-S1 修复：原 Android 端无任何 URLValidator 等价组件，OpenAIHttpTransport 与
 * WebSocketTransport 直接使用 config.serverUrl 拼接路径并发起请求，且在 Authorization
 * 头中携带 apiKey。攻击者控制 Agent 配置即可触发 SSRF。
 */
object UrlValidator {

    /** 允许的 scheme 白名单 */
    val allowedSchemes: Set<String> = setOf("http", "https", "ws", "wss")

    /**
     * 禁止访问的 IP 段（链路本地 / 元数据服务 / 回环陷阱等）。
     * 注意：127.0.0.1 / localhost 允许（本地 Agent 部署场景）
     */
    private val blockedCIDRs: List<Pair<ByteArray, Int>> = listOf(
        byteArrayOf(169.toByte(), 254.toByte(), 0.toByte(), 0.toByte()) to 16,  // 链路本地（含 AWS/GCP/Azure metadata）
        byteArrayOf(0.toByte(), 0.toByte(), 0.toByte(), 0.toByte()) to 8,       // 本网络
        byteArrayOf(255.toByte(), 255.toByte(), 255.toByte(), 255.toByte()) to 32, // 广播
        byteArrayOf(100.toByte(), 64.toByte(), 0.toByte(), 0.toByte()) to 10,   // CGNAT
        byteArrayOf(192.toByte(), 0.toByte(), 0.toByte(), 0.toByte()) to 24,    // IETF 协议分配
        byteArrayOf(198.toByte(), 18.toByte(), 0.toByte(), 0.toByte()) to 15,   // 基准测试
        byteArrayOf(224.toByte(), 0.toByte(), 0.toByte(), 0.toByte()) to 4,     // 组播
        byteArrayOf(240.toByte(), 0.toByte(), 0.toByte(), 0.toByte()) to 4      // 保留
    )

    /**
     * 校验 URL 字符串是否安全可用。
     *
     * @param urlString 用户输入的 URL 字符串
     * @param allowLocalhost 是否允许 localhost / 127.0.0.1 / 内网私网地址（默认 true，本地 Agent 部署需要）
     * @return 校验通过返回 URL；失败返回 null
     */
    fun validate(urlString: String, allowLocalhost: Boolean = true): URL? {
        val url = try { URL(urlString) } catch (_: Exception) { return null }
        val scheme = url.protocol?.lowercase() ?: return null
        if (scheme !in allowedSchemes) return null

        val host = url.host?.lowercase() ?: return null
        if (host.isEmpty()) return null

        // localhost 显式放行
        if (allowLocalhost && (host == "localhost" || host == "ip6-localhost" || host == "::1")) {
            return url
        }

        // IPv4 字面量校验
        parseIPv4(host)?.let { ipv4 ->
            return if (isIPv4Allowed(ipv4, allowLocalhost)) url else null
        }

        // IPv6 字面量
        if (host.contains(":")) {
            val lower = host.removeSurrounding("[", "]").lowercase()
            // 链路本地 fe80::/10（覆盖 fe80~febf）
            if (lower.startsWith("fe8") || lower.startsWith("fe9") ||
                lower.startsWith("fea") || lower.startsWith("feb")) {
                return null
            }
            // 唯一本地地址 fc00::/7
            if (lower.startsWith("fc") || lower.startsWith("fd")) {
                return if (allowLocalhost) url else null
            }
            // IPv4-mapped IPv6: ::ffff:a.b.c.d
            if (lower.startsWith("::ffff:")) {
                val ipv4Part = lower.removePrefix("::ffff:")
                parseIPv4(ipv4Part)?.let { ipv4 ->
                    return if (isIPv4Allowed(ipv4, allowLocalhost)) url else null
                }
            }
            return url
        }

        // 域名校验：禁止 .local / .internal（mDNS SSRF）
        if (host.endsWith(".local") || host.endsWith(".internal")) {
            return if (allowLocalhost) url else null
        }
        return url
    }

    // MARK: - IPv4 Helpers

    private fun parseIPv4(s: String): ByteArray? {
        val parts = s.split(".")
        if (parts.size != 4) return null
        val bytes = ByteArray(4)
        for ((i, p) in parts.withIndex()) {
            if (p.length > 3) return null
            val v = p.toIntOrNull() ?: return null
            if (v < 0 || v > 255) return null
            bytes[i] = v.toByte()
        }
        return bytes
    }

    private fun isIPv4Allowed(ip: ByteArray, allowLocalhost: Boolean): Boolean {
        // 127.0.0.0/8 回环
        if (ip[0] == 127.toByte()) return allowLocalhost
        // 私网地址
        if (ip[0] == 10.toByte()) return allowLocalhost
        if (ip[0] == 172.toByte() && ip[1] in 16..31) return allowLocalhost
        if (ip[0] == 192.toByte() && ip[1] == 168.toByte()) return allowLocalhost

        // 禁止的 IP 段
        for ((cidr, prefix) in blockedCIDRs) {
            if (inCIDR(ip, cidr, prefix)) return false
        }
        return true
    }

    private fun inCIDR(ip: ByteArray, cidr: ByteArray, prefix: Int): Boolean {
        val fullBytes = prefix / 8
        val remainder = prefix % 8
        for (i in 0 until fullBytes) {
            if (ip[i] != cidr[i]) return false
        }
        if (remainder > 0 && fullBytes < 4) {
            val mask: Byte = ((0xFF shl (8 - remainder)) and 0xFF).toByte()
            if ((ip[fullBytes].toInt() and mask.toInt()) != (cidr[fullBytes].toInt() and mask.toInt())) {
                return false
            }
        }
        return true
    }
}
